File: boj/28. DFS와BFS/2178_미로탐색.cpp
// bfs로 풀이
#include <iostream>
#include <algorithm>
#include <queue>
using namespace std;

int graph[100][100];
int n,m;
void bfs(int x, int y){
	int dx[]={0,0,1,-1};
	int dy[]={-1,1,0,0};
	queue<pair<int,int>> q;
	q.push(make_pair(0,0));
	pair<int,int> cur;
	
	while(!q.empty()){
		cur=q.front();
		q.pop();
		
		for(int i=0;i<4;i++){
			int nx=cur.second+dx[i];
			int ny=cur.first+dy[i];
			if(nx>=0&&ny>=0&&nx<m&&ny<n&&graph[ny][nx]==1){
				q.push(make_pair(ny,nx));
				graph[ny][nx]=graph[cur.first][cur.second]+1;
			}
		}
	}
}
int main() {
	string s;
	cin>>n>>m;
	
	for(int i=0;i<n;i++){
		cin>>s;
		for(int j=0;j<m;j++){
			if(s[j]=='0')
				graph[i][j]=0;
			else
				graph[i][j]=1;
		}
	}
	bfs(0,0);
	cout<<graph[n-1][m-1]<<"\n";
	return 0;
}

File: boj/28. DFS와BFS/2178_미로탐색.py
from collections import deque

n, m = map(int, input().split())
graph = []

for _ in range(n):
	graph.append(list(map(int, input().rstrip())))

def bfs(x, y):
	queue = deque()
	queue.append((x,y))
	
	dx = [-1, 0, 0, 1]
	dy = [0, -1, 1, 0]
	
	while queue:
		x, y = queue.popleft()
		
		for i in range(4):
			nx = x + dx[i]
			ny = y + dy[i]
			
			if 0 <= nx < n and 0 <= ny < m and graph[nx][ny] == 1:
				graph[nx][ny] = graph[x][y] + 1 # 이전 노드에서 현재 노드로 이동 시, (이전 노드 + 1) 값 저장 -> 방문한 칸 수. 시작 노드 1부터.
				queue.append((nx,ny))
# for g in graph:
# 	print(g)
bfs(0,0)
# for g in graph:
# 	print(g)
print(graph[-1][-1])